### PR TITLE
Rent vs mortgage maps

### PR DESCRIPTION
**Pull request by elyobo**
Thanks to @MaxwellBo for his assistance getting this started.

Any review is welcome, including code review and suggestions on what / how to display things etc.

Visualises the geographic distribution of the rent vs buy ratio.

image

### CODE REVIEW

File: src/rent-vs-mortgage.md [hunk 1]
```diff
@@ -362,44 +356,62 @@ Inputs.table(search, {
 </div>
 
 
-## Where are the mortages "affordable" (closest to rents)?
+## Where are the mortages relative "affordable" (closest to rents)?
```
Comment (MaxwellBo) [inline src/rent-vs-mortgage.md:359]:
```suggestion
## Where are the mortgages relatively "affordable" (closest to rents)?
```

File: src/rent-vs-mortgage.md [hunk 2]
```diff
@@ -362,44 +356,62 @@ Inputs.table(search, {
...
 
-<div class="card" style="margin: 0 -1rem;">
+```js
+const propertyType = view(Inputs.radio(['HOUSE', 'UNIT'], { label: 'Property Type', value: 'HOUSE' }))
+const rangeType = view(Inputs.radio(['DYNAMIC', 'FIXED'], { label: 'Colour Range Type', value: 'FIXED' }))
+```
+
+The FIXED colour range option ensures the same fixed range of colours for consistent display across both units and houses, while the DYNAMIC option adjusts the range based on the data to maximise the contrast.
```
Comment (MaxwellBo) [inline src/rent-vs-mortgage.md:366]:
```suggestion
The `FIXED` colour range option ensures the same fixed range of colours for consistent display across both units and houses, while the `DYNAMIC` option adjusts the range based on the data to maximise the contrast.
```

File: src/rent-vs-mortgage.md [hunk 3]
```diff
@@ -362,44 +356,62 @@ Inputs.table(search, {
 </div>
 
 
-## Where are the mortages "affordable" (closest to rents)?
+## Where are the mortages relative "affordable" (closest to rents)?
 
-<div class="card" style="margin: 0 -1rem;">
+```js
+const propertyType = view(Inputs.radio(['HOUSE', 'UNIT'], { label: 'Property Type', value: 'HOUSE' }))
+const rangeType = view(Inputs.radio(['DYNAMIC', 'FIXED'], { label: 'Colour Range Type', value: 'FIXED' }))
```
Comment (MaxwellBo) [inline src/rent-vs-mortgage.md:363]:
```suggestion
const propertyType = view(Inputs.radio(['HOUSE', 'UNIT'], { label: 'Property type', value: 'HOUSE' }))
const rangeType = view(Inputs.radio(['DYNAMIC', 'FIXED'], { label: 'Colour range type', value: 'FIXED' }))
```

File: src/rent-vs-mortgage.md [hunk 4]
```diff
@@ -362,44 +356,62 @@ Inputs.table(search, {
...
+const sa2Map = new Map(dataArray.map(row => [row.sa2_code.toString(), row]))
+const sa2Codes = new Set(sa2Map.keys())
+const domain = rangeType === 'FIXED'
+   ? FIXED_DOMAIN
+   : dataArray.reduce((acc, row) => {
+        const ratio = row.repayment_to_rent_ratio
+        return [Math.min(acc[0], ratio), Math.max(acc[1], ratio)]
+      },
+      [100, 0],
+   )
```
Comment (MaxwellBo) [inline src/rent-vs-mortgage.md:379]: shouldn't Plot dynamically compute the domain for the color scale here? We could yield `undefined` on the else condition if that's the case

File: src/rent-vs-mortgage.md [hunk 5]
```diff
@@ -362,44 +356,62 @@ Inputs.table(search, {
...
+
+const deckColours = Plot.plot({ color: { domain } }).scale('color')
+const plotRatioLegend = Plot.legend({
+  label: 'Repayment to rent ratio',
+  color: deckColours,
+})
+```
```
Comment (MaxwellBo) [inline src/rent-vs-mortgage.md:386]:
```suggestion
const deckColors = Plot.plot({ color: { domain } }).scale('color')
const plotRatioLegend = Plot.legend({
  label: 'Repayment to rent ratio',
  color: deckColors,
})
```
```

File: src/rent-vs-mortgage.md [hunk 6]
```diff
@@ -432,73 +444,70 @@ const deckInstance = new DeckGL({
...
-      // getLineColor: () => [255, 255, 255],
+      getLineColor: 0,
+      getLineWidth: 20,
     }),
   ],
-  getTooltip: ({object}) => {
-    return object && object.properties.SA2_MAIN16
-  } 
-});
+  getTooltip: (point) => {
```
Comment (MaxwellBo) [inline src/rent-vs-mortgage.md:467]: ah awesome, I wasn't quite sure why this wasn't working

File: src/rent-vs-mortgage.md [hunk 7]
```diff
@@ -432,73 +444,70 @@ const deckInstance = new DeckGL({
...
-          d.sa2_code.toString() === f.properties.SA2_MAIN16
-        );
-
-        if (!sa2) {
-          return undefined;
-        }
-
-        return sa2.repayment_to_rent_ratio
-      },
+      fill: f => sa2Map.get(f?.properties?.REGION_CODE)?.repayment_to_rent_ratio,
```
Comment (MaxwellBo) [inline src/rent-vs-mortgage.md:498]: we might want to add a `stroke: "black"` to this if want the Deck and Plot maps to look the same

**Review by MaxwellBo — CHANGES_REQUESTED** [review]
left some comments

File: src/rent-vs-mortgage.md (hunk 4 above)
Comment (elyobo) [reply src/rent-vs-mortgage.md:379]: Oh, that's a great point - definitely easier to leave it to Plot in that case, I'll check it out.

File: src/rent-vs-mortgage.md (hunk 5 above)
Comment (elyobo) [reply src/rent-vs-mortgage.md:386]: I refuse to give in to American cultural hegemony :D

File: src/rent-vs-mortgage.md (hunk 7 above)
Comment (elyobo) [reply src/rent-vs-mortgage.md:498]: Yeah, better like this for my tastes.

image

File: src/rent-vs-mortgage.md (hunk 4 above)
Comment (elyobo) [reply src/rent-vs-mortgage.md:379]: Yeah that works, just need to actually plug in some data into the unrendered plot below that I use to get the range.

**Comment by elyobo** [comment]
Thanks @MaxwellBo, feedback addressed - aside from the Americanisation!

**Review by MaxwellBo — APPROVED** [review]

**Pull request merged**